subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/dependencysubstitution/DependencySubstitutionResolver.java

```diff
@@ -36,11 +37,13 @@ public DependencySubstitutionResolver(DependencyToComponentIdResolver resolver,
     public void resolve(DependencyMetadata dependency, BuildableComponentIdResolveResult result) {
         ComponentSelector selector = dependency.getSelector();
         DependencySubstitutionInternal details = new DefaultDependencySubstitution(selector, dependency.getRequested());
-        try {
-            rule.execute(details);
-        } catch (Throwable e) {
-            result.failed(new ModuleVersionResolveException(selector, e));
-            return;
+        synchronized (DependencySubstitutionResolver.class) {
```

Comment:
@melix Wouldn't the same problem apply to component metadata rules etc.? Or are they run at a different point of the lifecycle?